packages/jsii-pacmak/lib/targets/golang/module.ts

```diff
@@ -1,64 +1,254 @@
 import { CodeMaker } from 'codemaker';
-import { Assembly, Submodule as JsiiSubmodule } from 'jsii-reflect';
+import { Assembly, Type, Submodule as JsiiSubmodule } from 'jsii-reflect';
+import { join } from 'path';
 import { GoClass, Enum, Interface } from './types';
 
-export interface ModuleTypes {
-  [fqn: string]: Interface | Enum | GoClass;
+type ModuleType = Interface | Enum | GoClass;
+type ModuleTypes = ModuleType[];
+
+// TODO: Make this a class?
+export interface Module {
+  root: RootModule;
+  packageName: string;
+  moduleName: string;
+  filePath: string;
+  file: string;
+  submodules: Submodule[];
+  types: ModuleTypes;
+  hasSubmodules(): boolean;
+}
+
+function buildSubmodules(
+  root: RootModule,
+  parent: Module,
+  submodules: readonly JsiiSubmodule[],
+): Submodule[] {
+  return submodules.map((sm) => new Submodule(root, parent, sm));
 }
 
-export class Module {
-  public readonly types: ModuleTypes = {};
-  public readonly submodules: Submodule[];
-  public readonly assembly: Assembly | JsiiSubmodule;
+function unknownType(type: Type): never {
+  throw new Error(`Type: ${type.name} is not an interface, enum, or class`);
+}
 
-  public constructor(assembly: Assembly, submodule?: JsiiSubmodule) {
-    this.assembly = submodule ?? assembly;
-    assembly.types.forEach((type) => {
-      let t: Enum | Interface | GoClass | undefined;
+function buildModuleTypes(parent: Module, types: readonly Type[]): ModuleTypes {
+  return types.map(
+    (type: Type): ModuleType => {
       if (type.isInterfaceType()) {
-        t = new Interface(type);
+        return new Interface(parent, type);
       } else if (type.isClassType()) {
-        t = new GoClass(type);
+        return new GoClass(parent, type);
       } else if (type.isEnumType()) {
-        t = new Enum(type);
+        return new Enum(parent, type);
       }
+      return unknownType(type);
+    },
+  );
+}
 
-      if (t) {
-        this.types[type.fqn] = t;
-      }
-    });
+function findTypeInTree(module: Module, fqn: string): ModuleType | undefined {
+  const result = module.types.find((t) => t.type.fqn === fqn);
 
-    this.submodules = this.buildSubmodules(assembly, submodule);
+  if (result) {
+    return result;
+  } else if (module.hasSubmodules()) {
+    return module.submodules.reduce((accum: ModuleType | undefined, sm) => {
```

Comment:
If `hasSubmodules()` is false, would `submodules` be an empty collection? If so, the nesting/if is unnecessary.